src/util/transaction_identifier.h

```diff
@@ -0,0 +1,61 @@
+#ifndef BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
+#define BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
+
+#include <uint256.h>
+#include <util/types.h>
+
+#include <type_traits>
+
+/** transaction_identifier represents the two canonical transaction identifier
+ * types (txid, wtxid).
+ *
+ * TODO: This type is not fully type-safe yet as we allow implicit conversion
+ * to uint256. We should make it impossible to implicitly convert from
+ * transaction_identifier to uint256 (e.g. by making the base class private).
+ * This should be easy to do once most of the code base has converted to the
+ * Txid and Wtxid types to avoid churn. */
+template <bool has_witness>
+class transaction_identifier : public uint256
+{
+public:
+    transaction_identifier() : uint256{} {}
+
+    template <typename Other>
+    transaction_identifier(const Other& other)
+        : uint256{other}
+    {
+        static_assert(std::is_same_v<Other, transaction_identifier<has_witness>>,
+                      "Forbidden copy type");
+    }
```

Comment:
This method seems to be unused -- remove for now?

(Perhaps `FromUint256()` has taken its place?)

Reply:
I would like to keep this to make the allowed copy types explicit. This can be removed once we disallow `uint256 a = Wtxid/Txid{}`.